File: 26-Visitor/src/go/addition.go
package main

// Addition is a Concrete Element — addition of two sub-expressions.
type Addition struct {
	left  Expression
	right Expression
}

func NewAddition(left, right Expression) *Addition {
	return &Addition{left: left, right: right}
}

func (a *Addition) Left() Expression  { return a.left }
func (a *Addition) Right() Expression { return a.right }

func (a *Addition) Accept(visitor Visitor) {
	visitor.VisitAddition(a)
}

File: 26-Visitor/src/rust/addition.rs
use crate::expression::Expression;
use crate::visitor::Visitor;

/// Concrete Element — addition of two sub-expressions (owns its children).
pub struct Addition {
    left: Box<dyn Expression>,
    right: Box<dyn Expression>,
}

impl Addition {
    pub fn new(left: Box<dyn Expression>, right: Box<dyn Expression>) -> Self {
        Addition { left, right }
    }

    pub fn left(&self) -> &dyn Expression {
        self.left.as_ref()
    }

    pub fn right(&self) -> &dyn Expression {
        self.right.as_ref()
    }
}

impl Expression for Addition {
    fn accept(&self, visitor: &mut dyn Visitor) {
        visitor.visit_addition(self);
    }
}
